### PR TITLE
User sign in

### PR DESCRIPTION
**Pull request by shernanm**
## Summary

* The login was implemented.
* Authentication by means of token was implemented.

## Trello Card

https://trello.com/c/XN5atBMq

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -1,16 +1,41 @@
 const validations = require('../util');
```
Comment (molinajulian) [inline app/controllers/user.js]: Rename for `utils`. Remember, you must be declarative. Also, the file must be `utils.js`

File: app/errors.js [hunk 2]
```diff
@@ -16,5 +16,16 @@ exports.BAD_REQUEST = 'bad_request';
...
+exports.PASSWORD_DECRYPTION_ERROR = 'password_decryption_error.';
+exports.decryptionError = () =>
+  internalError('Error trying to decrypt the password.', exports.PASSWORD_DECRYPTION_ERROR);
+
+exports.SIGN_UP_ERROR = 'sign_up_error';
+exports.signUpError = message => internalError(message, exports.SIGN_UP_ERROR);
+
+exports.GENERATE_TOKEN_ERROR = 'generate_token_error.';
+exports.generateTokenError = () =>
+  internalError('Error trying to generate the token.', exports.GENERATE_TOKEN_ERROR);
```
Comment (molinajulian) [inline app/errors.js:39]: Move internal codes to top, and methods for bottom or on the contrary :+1: . This form is more organized.

File: app/controllers/user.js [hunk 3]
```diff
@@ -1,16 +1,41 @@
...
     })
     .catch(next);
+};
+
+exports.signInUser = (req, res, next) => {
+  logger.info('SignInUser method start.');
+  const { email, password } = req.body;
+  return userService
+    .findUser(email)
+    .then(foundUser => {
```
Comment (molinajulian) [inline app/controllers/user.js]: This works but I think that you can improve this. Specially, the part when you return null. What should happen if user does not exist?

File: app/controllers/user.js [hunk 4]
```diff
@@ -1,16 +1,41 @@
...
+      if (foundUser) {
+        return validations
+          .passwordDecryption(password, foundUser.password)
```
Comment (molinajulian) [inline app/controllers/user.js]: It is not bad that you save this method in utils, but you should know that you can create a service for `jwt`. Check the advantages and disadvantages of these two options and then we talk

File: app/services/users.js [hunk 5]
```diff
@@ -7,3 +7,13 @@ exports.userRegister = user =>
     logger.info(`Error trying to create the user ${user.firstName} ${user.lastName}`);
     throw errors.databaseError('Error processing request in database.');
   });
+
+exports.findUser = email =>
```
Comment (molinajulian) [inline app/services/users.js]: Maybe create `findUserBy` and you pass options. Its more scalable

File: app/util.js [hunk 6]
```diff
@@ -10,6 +11,16 @@ exports.passwordEncryption = user =>
       ...user,
       password
     }))
-    .catch(err => {
-      throw errors.encryptionError(`${err.message}`);
+    .catch(() => {
```
Comment (molinajulian) [inline app/util.js]: You can do `.catch(errors.encryptionError)`. You pass implicitly the parameters received by catch and return the function. Its similar to do `.catch(err => errors.encryptionError(err))`.

File: test/users.test.js [hunk 7]
```diff
@@ -100,3 +102,38 @@ describe('POST /users', () => {
         dictum.chai(response, 'Test when required parameters are not sent');
       }));
 });
+
+describe('POST /users/sessions', () => {
+  test('Successful test when sign in', () =>
```
Comment (molinajulian) [inline test/users.test.js:113]: Add more cases

File: test/users.test.js [hunk 8]
```diff
@@ -100,3 +102,38 @@ describe('POST /users', () => {
...
+      })
+      .then(() =>
+        controller
+          .post('/users/sessions')
+          .send({
+            email: 'john.katz@wolox.co',
+            password: '12345678'
+          })
+          .then(userSignIn =>
+            jwt.verifyAsync(userSignIn.body.token, config.seed).then(token => ({
```
Comment (molinajulian) [inline test/users.test.js]: Maybe create a function to check token in tests?

File: test/users.test.js [hunk 9]
```diff
@@ -100,3 +102,38 @@ describe('POST /users', () => {
         dictum.chai(response, 'Test when required parameters are not sent');
       }));
 });
+
+describe('POST /users/sessions', () => {
+  test('Successful test when sign in', () =>
+    controller
+      .post('/users')
+      .send({
```
Comment (molinajulian) [inline test/users.test.js]: Maybe create a const for this object

**Review by molinajulian — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 10]
```diff
@@ -6,7 +6,7 @@ node_js:
 sudo: true
 
 env:
-  - CXX=g++-4.8 NODE_ENV=testing
+  - CXX=g++-4.8 NODE_ENV=testing SEED=wolox-training
```
Comment (raparicio6) [inline .travis.yml]: where it is defined this seed you are using?

File: app/services/users.js [hunk 11]
```diff
@@ -1,9 +1,38 @@
...
 exports.userRegister = user =>
   User.create(user).catch(() => {
     logger.info(`Error trying to create the user ${user.firstName} ${user.lastName}`);
     throw errors.databaseError('Error processing request in database.');
   });
+
+exports.findUserBy = option =>
+  User.findOne({
+    where: option,
+    raw: true,
```
Comment (molinajulian) [inline app/services/users.js:15]: Why ?

**Review by molinajulian — APPROVED** [review]

**Review by raparicio6 — APPROVED** [review]

**Pull request merged**